clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -295,24 +297,27 @@ private long sendProducerData(long now) {
     }
 
     private boolean maybeSendTransactionalRequest(long now) {
-        TransactionManager.TxnRequestHandler nextRequestHandler = transactionManager.nextRequestHandler();
-        if (nextRequestHandler == null) {
-            log.trace("TransactionalId: {} -- There are no pending transactional requests to send",
-                    transactionManager.transactionalId());
-            return false;
-        }
+        String transactionalId = transactionManager.transactionalId();
+        if (transactionManager.isCompletingTransaction() && !transactionManager.hasPartitionsToAdd() && accumulator.hasUnflushedBatches()) {
```

Comment:
Note that this is a bug fix. The call to `nextRequestHandler` may complete the transaction without sending an EndTxnRequest. In that case, there would still be queued produce requests which will probably fall into the next transaction.